[WIP] feat(pubsub): Publishing of Events

florianfischeer opened the pull request:
This pull request adds the publishing of events to the PubSub part of the stack.
We added a cmake-flag, which is called `UA_ENABLE_PUBSUB_EVENTS`, a tutorial for publishing and another one for subscribing.

## Concept
In the PubSubManager we added a list that saves all PublishedDataSets of the type `UA_PUBSUB_DATASET_PUBLISHEDEVENTS` with their associated DataSetWriters. Besides, a DataSetWriter now contains a queue in which the information of the events is saved and then sent after the publishing interval has expired. When creating a DataSetWriter, if the associated PublishedDataSet is of the mentioned type, both are added to the PubSubManager list. This list helps us to recognize when an event has occurred whether it belongs to a PublishedDataSet. If so, the DataSetFields are added to the queue of the DataSetWriter belonging to the PublishedDataSet.

## Specific Changes
### Added structs
To implement our concept, we had to create some new structs. First, we needed a `UA_DataSetEventConfig`, because there was only a placeholder in the UA_DataSetFieldConfig. From our point of view, an EventDataSetField contains a selectedField, a fieldname and a boolean, which shows, whether it's a promoted field. Also, we added a pointer of selectedFields to the UA_PublishedEventConfig.

Moreover, we needed some structs to implement the Queue in the DataSetWriter and the list in the PubSubManager. The Queue in the DataSetWriter consists of `EventQueueEntries`, which are only a wrapper for DataValues. The list of the PubSubManager on the other hand consists of `PublishedDataSetEventEntries`. This struct contains a field for a DataSetWriter and one for its connected PublishedDataSet.

### Changed Methods
For our implementation, the most important method is `UA_Server_triggerEvent`, because it's necessary to call it if you want to trigger an event. In this method, we check, whether the triggered Event matches with a PublishedDataSet and if so we call `addEventToDataSetWriter`. This method is new and it iterates over all DataSetFields of the given PublishedDataSet, wraps them into a DataValue and passes it to the `insertDataValueIntoDSWQueue`-method. In here the passed DataValue is added to the Queue in the passed DataSetWriter.

If the publishing interval expired and a KeyFrameMessage needs to be created, we fill the DataSetMessage with the DataValues in the Queue. After this, the Queue is cleared and all information is published.

## Additions to the Informationmodel
We added functionality to support PublishedDataSets of the PublishedEventsType in the information model. Most of the functionality is implemented analogue to the PublishedDataItemsType. 

### Added Methods
The first method we added is `addPublishedEventsRepresentation` which is called every time a PDS is created by `UA_Server_addPublishedDataSet`, provided that the information model is enabled and the created PDS is of the PublishedEventsType. This method adds the required nodes to represent a PublishedEventsType Object to the information model.

We also implemented the `addPublishedEvents` method for the DataSetFolderType (9.1.4.5.3) which, when called, creates a PDS of PublishedEventsType.

### Added to the NodeSet files
To fully support the PublishedEventsType even when running at reduced ns0 we had to add a few nodes to the NodeSet files. More precisely we added `AddPublishedEvents` and `PublishedEventsType` to Opc.Ua.NodeSet2.PubSubMinimal.xml and dependent nodes to Opc.Ua.NodeSet2.Reduced.xml.


>There are a few points that we mention here that are currently in progress and not merged into this branch

florianfischeer committed:
feat(pubsub): Publish Events

Tym0r committed:
feat(pubsub): Events Informationmodel-Methods

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/open62541/open62541?pullRequest=4220) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/open62541/open62541?pullRequest=4220) before we can accept your contribution.<br/>**0** out of **2** committers have signed the CLA.<br/><br/>:x: Florianfischer42<br/>:x: Tym0r<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/open62541/open62541?pullRequest=4220) it.</sub>

jpfr commented:
Very good feature and also code quality from a first look.

@andreasebner can we green-light the public API before doing an in-depth review?

Florianfischer42 commented:
Thanks for your comment on our pull request. After a meeting with @andreasebner we recognized, that there still need to be made some adjustments to match the feature with the specification.

andreasebner commented:
> Very good feature and also code quality from a first look.
> 
> @andreasebner can we green-light the public API before doing an in-depth review?

Yes, for sure, I will trigger you if we reach the final stage. The public API should not be changed a lot since events in PubSub are "only" another type of PublishedDataSets.

andreasebner commented:
@Florianfischer42 Please provide a description for this PR. The description should explain your work, refer to the relevant PubSub specification sections and list changes/additions on the public API.

florianfischeer committed:
test(pubsub): Publish Events

florianfischeer committed:
feat(pubsub): FieldMetaData for Event-DSF

Tym0r committed:
feat(pubsub): Evaluation of Event Contentfilter

Tym0r committed:
feat(pubsub): Events in config-files

jpfr commented:
Looking good! Two quick questions:

- Is the eventQueue protected against overflow? That is, if events are generated faster than we can send them out, where will things break?
- You manage a global list of `PublishedDataSetEventEntry`.
  - Iterating over the global list for all events could become inefficient when the number of PublishedDataSetEventEntry gets large.
  - Does that cover Events that "blubble up" in the object hierarchy?
You might want to reuse functionality from the Event-MonitoredItems for that.
They are added to a linked-list of the node where they "collect" events bubbling up from below objects in the hierarchy.

Keep up the good work!

Tym0r committed:
changes from code review

- bubbling-up
- moves eventQueueMaxSize to config
- removes reference to unimplemented ModifyFieldSelection
- adds comments for the global event-PDS list

Florianfischer42 commented:
@jpfr 
Thank you for the feedback.
### Is the eventQueue protected against overflow? That is if events are generated faster than we can send them out, where will things break?
We added a basic overflow protection, which means if the queue is full or memory could not be allocated it will just fail to add more events and return an error code.
### Iterating over the global list could become inefficient
For now, a global list that contains event-only PDS is the most efficient way we could come up with. If you can suggest a better solution, please let us know.
### Are events bubbling up in the object hierarchy?
That's a good point, bubbling up is now supported.

andreasebner commented on src/pubsub/ua_pubsub_ns0.c:
```diff
@@ -595,6 +648,108 @@ removeDataSetFolderAction(UA_Server *server,
+                                     filterContext);
+
+#ifdef UA_ENABLE_PUBSUB_INFORMATIONMODEL_METHODS
+    retVal |= UA_Server_addReference(server, publishedDataSet->identifier,
+                                     UA_NODEID_NUMERIC(0, UA_NS0ID_HASCOMPONENT),
+                                     UA_EXPANDEDNODEID_NUMERIC(0, UA_NS0ID_PUBLISHEDEVENTSTYPE_MODIFYFIELDSELECTION), true);
```
I think you currently did not add the implementation of the 'MODIFYFIELDSELECTION' function, right?

Tym0r replied:
Right, it's still on the todo list. I commented it out with a todo notice. Probably makes sense to change the event config from DSF to PDS before implementing this method.

andreasebner commented on src/pubsub/ua_pubsub_writer.c:
```diff
@@ -697,9 +816,18 @@ UA_Server_addDataSetField(UA_Server *server, const UA_NodeId publishedDataSet,
+    } else if(newField->config.dataSetFieldType == UA_PUBSUB_DATASETFIELD_EVENT){
+        if(newField->config.field.events.promotedField)
+            currentDataSet->promotedFieldsCount++;
+        currentDataSet->fieldSize++;
```
Is it necessary to call addDataSetField after creating an Event-PDS?

Tym0r replied:
Currently yes, it sets the correct PDS fieldSize, the promotedFields and die FieldMetaData. It wouldn't be too difficult to rewrite it though.

andreasebner commented on src/pubsub/ua_pubsub_writer.c:
```diff
@@ -1288,6 +1417,22 @@ UA_Server_addDataSetWriter(UA_Server *server,
+#ifdef UA_ENABLE_PUBSUB_EVENTS
+    if(currentDataSetContext->config.publishedDataSetType == UA_PUBSUB_DATASET_PUBLISHEDEVENTS){
+        SIMPLEQ_INIT(&newDataSetWriter->eventQueue); // only needs to be initalized if the linked publishedDataSet is for event
+        newDataSetWriter->eventQueueMaxSize = 1000; //test value
```
This parameter should be more visible.

andreasebner commented on src/server/ua_subscription_events.c:
```diff
@@ -519,6 +521,60 @@ static const UA_NodeId isInFolderReferences[2] =
+static UA_StatusCode
+insertDataValueIntoDSWQueue(UA_Server *server, UA_DataSetWriter *dsw, UA_DataValue *value)  {
+    if(dsw->eventQueueEntries == dsw->eventQueueMaxSize){
+        UA_LOG_ERROR(&server->config.logger, UA_LOGCATEGORY_SERVER,
+                     "EventQueueMaxSize reached %lu / %lu", dsw->eventQueueEntries, dsw->eventQueueMaxSize);
```
I think a warning is enough. Just drop the new field or remove the oldest entry and add this new one.

andreasebner reviewed (commented):
Thanks’ for your great contribution!

I've added some open questions and mentioned some problems/formatting issues. In addition please clean up your commit history.

Florianfischer42 commented on src/pubsub/ua_pubsub_manager.h:
```diff
@@ -16,12 +16,24 @@ _UA_BEGIN_DECLS
 
+#ifdef UA_ENABLE_PUBSUB_EVENTS
+typedef struct PublishedDataSetEventEntry {
+    UA_PublishedDataSet *pds;
```
It isn't needed. We added it, to improve runtime. The PublishedDataSet can't be accessed directly from a DSW, it only offers the NodeId of a PDS. The algorithm to find the PDS (UA_PublishedDataSet_findPDSbyId) iterates over all PDSs, this would affect the efficiency.

Tym0r commented on examples/pubsub/tutorial_pubsub_publish_event.c:
```diff
@@ -0,0 +1,327 @@
+static void addDataSetFields(UA_Server *server){
+    /*Creates DataSetFields based on the PDS selectedFields*/
+    UA_PublishedDataSet *publishedDataSet = UA_PublishedDataSet_findPDSbyId(server, publishedDataSetIdent);
+    for(size_t i = 0; i < publishedDataSet->config.config.event.selectedFieldsSize; i++){
+        UA_SimpleAttributeOperand sao = publishedDataSet->config.config.event.selectedFields[i];
+        UA_String name = sao.browsePath[0].name;
+
+        UA_NodeId dataSetFieldIdent;
+        UA_DataSetFieldConfig dataSetFieldConfig;
+        memset(&dataSetFieldConfig, 0, sizeof(UA_DataSetFieldConfig));
+        dataSetFieldConfig.dataSetFieldType = UA_PUBSUB_DATASETFIELD_EVENT;
+        dataSetFieldConfig.field.events.fieldNameAlias = name;
```
The fieldNameAlias is used in the FieldMetaData and the AddPublishedEvents method has a FieldNameAliases parameter. I agree that we probably don't need the DSF-config at all, we could just add the two fields to the PDS-config. We started out with the DSF-config because we didn't know the exact config that was needed.

Tym0r committed:
fixes problem with updated transport profile

AlexexanderSavinov commented:
This one is still open as i understand.
I am currently on master branch and use PubSub and need this changes.
Is there a plan when they are going to be included?

CodingLappen commented:
I also like to know when this commit will be merged or if it will be merged.

jpfr commented:
@Florianfischer42 Are you still on this?
There are a few review comments from @andreasebner still unaddressed.

applea9 commented:
I am also interested in this PR, is there any plan to merge this PR to master?

wlkrm commented:
I interested in this PR, too.  Is there any plan to merge this PR to master? If you have 
no plans in continuing this PR, I might address the unaddressed review comments, @florianfischeer ?

jpfr commented:
As @florianfischeer is not responsing, I think you can take this up @wlkrm !

CodingLappen commented:
I have a working version already completed, I just didnt create a pull request , because I have clean up the code (more or less), but I have to clean up the commits, just didn't find the time to do this yet.

jpfr commented:
Converting to draft because of the WIP title.
